opencti-platform/opencti-graphql/src/migrations/1712225747215-convert_linked-to_related-to.js

```diff
@@ -0,0 +1,128 @@
+    `;
+  const relToParentTypes = ['basic-relationship', 'stix-relationship', 'stix-core-relationship'];
+
+  const reindexLinkedToToRelatedToQuery = {
+    source: {
+      index: INDEX_STIX_META_RELATIONSHIPS,
+      query: {
+        term: { 'entity_type.keyword': { value: linkedToType } }
+      }
+    },
+    dest: {
+      index: INDEX_STIX_CORE_RELATIONSHIPS
+    },
+    script: {
+      source: reindexLinkedToToRelatedToSource,
+      params: { linkedToType, relToType: RELATION_RELATED_TO, relToParentTypes, relatedToRelationsFromToArray }
+    }
+  };
+
+  await elReindexByQueryForMigration('[MIGRATION] Reindexing and updating non-duplicate linked-to refs', null, reindexLinkedToToRelatedToQuery);
+
+  // Then, We need to update all rel that had a linked-to ref as from or to
+  const relToTypes = [RELATION_RELATED_TO, 'basic-relationship', 'stix-relationship', 'stix-core-relationship'];
+  const updateRelWithLinkedToFromOrToSource = `
+    if(ctx._source.fromType == params.linkedToType) {ctx._source.fromType=params.relToType}
+    if(ctx._source.toType == params.linkedToType) {ctx._source.toType=params.relToType}
+    for(connection in ctx._source.connections) {
+      if(connection.types.contains(params.linkedToType)) {
+        connection.types = params.relToTypes;
+        connection.name = connection.name.replace("relationship-meta--", "relationship--");
+      }
+    }`;
+  const updateRelWithLinkedToFromOrToQuery = {
+    script: {
+      source: updateRelWithLinkedToFromOrToSource,
+      params: { linkedToType, relToType: RELATION_RELATED_TO, relToTypes }
+    },
+    query: {
+      nested: {
+        path: 'connections',
+        query: {
+          bool: {
+            must: [
+              { term: { 'connections.types.keyword': { value: linkedToType } } }
+            ]
+          }
+        },
+      }
+    }
+  };
+
+  await elUpdateByQueryForMigration('[MIGRATION] Updating relations with a linked-to from or to', [READ_RELATIONSHIPS_INDICES], updateRelWithLinkedToFromOrToQuery);
+
+  // // Then we need to move all denormalized linked-to rel in objects to related-to (if id is not already in related-to)
+  const updateDenormalizedLinkedToSource = `
+    ArrayList linkedTos = ctx._source[params.relLinkedTo];
+    if(!ctx._source.containsKey(params.relRelatedTo)) {ctx._source[params.relRelatedTo]=[] }
+    for(linkedTo in linkedTos) {
+      if(!ctx._source[params.relRelatedTo].contains(linkedTo)) { ctx._source[params.relRelatedTo].add(linkedTo) }
+    }
+    ctx._source.remove(params.relLinkedTo)`;
+  const updateDenormalizedLinkedToQuery = {
+    script: {
+      source: updateDenormalizedLinkedToSource,
+      params: { relLinkedTo, relRelatedTo }
+    },
+    query: {
+      exists: {
+        field: relLinkedTo
+      }
+    },
+  };
+
+  await elUpdateByQueryForMigration('[MIGRATION] Updating entities with rel_linked-to to rel_related-to', [READ_DATA_INDICES], updateDenormalizedLinkedToQuery);
```

Comment:
READ_DATA_INDICES includes all our indices with internal objects and relationships, but here I think we only need to read indices with stix objects right ? can we have linked-to refs between relationships ?

Reply:
we can't right now, it's only defined for stix objects you're right. I wasn't sure if it was possible in the past though, since having it between relationships could make sense. Is it too costly performance-wise to keep READ_DATA_INDICES just in case?